### PR TITLE
allow parallel class loading for the scala launcher

### PR DESCRIPTION
**Pull request by mkeskells**
Noted that the current Scala launcher is not registered as parallel capable. This means that there is a contention on class loading in multiple threads

Not sure which of the other scala defined classloaders should also be registered

If adopted, this would probably need a community build and mentioning in the release notes

### CODE REVIEW

**Comment by retronym** [comment]
Here's a [technote](https://docs.oracle.com/javase/7/docs/technotes/guides/lang/cl-mt.html) with that describes the change in JDK 1.7 that allowed finer grained locking during classloading. 

The stated motivation was to avoid deadlocks. I can see that reducing contention could also be a possible advantage, although I suspect that application profiles that point to multi-threaded contention problems in classloading could often suggest the need to re-use classloader instances (as we have done in the the compiler for compiler plugin and macro classloaders). I do understand Ihis isn't always possible e.g. if each run of some dynamically loaded script needs a truly clean slate of java-static state.

@mkeskells Did you come to this PR from a practical application performance issue or by discovering the problem via a review  of this code?

In any case, marking this laoder as parallel capable seems safe. Marking this loader as parallel capable is necessary but not sufficient condition for subclasses of it to be considered capable -- they need to also mark themselves as parallel capable.

One problem I see with the implementation is that the approximation of the static initializer doesn't quite get the order correct. A subclass of `ScalaClassLoader.URLClassLoader` that attempts to register itself as parallel capable would have hard time guaranteeing that `ScalaClassLoader.URLClassLoader` had first been registered. It may be simpler to deprecate `ScalaClassLoader.URLClassLoader` and provide a replacement (which could be used internally by Toolbox evaluation etc) that is implemented as a Java class.

**Comment by som-snytt** [comment]
Relatedly https://github.com/scala/bug/issues/11429

File: src/reflect/scala/reflect/internal/util/ScalaClassLoader.scala [hunk 1]
```diff
@@ -129,10 +129,17 @@ object ScalaClassLoader {
...
+  object URLClassLoader {
+    private val registered = new AtomicBoolean
+  }
   class URLClassLoader(urls: Seq[URL], parent: JClassLoader)
       extends JURLClassLoader(urls.toArray, parent)
          with ScalaClassLoader
          with HasClassPath {
+    // Note - this is calling a caller sensitive method so must be called from the class that is being registered
+    // and we are behaving like a java static method
+    if (URLClassLoader.registered.compareAndSet(false, true))
```
Comment (dwijnand) [inline src/reflect/scala/reflect/internal/util/ScalaClassLoader.scala]:
```suggestion
    if (!URLClassLoader.registered.getAndSet(true))
```

Even better as it works better on those new Apple M1 silicon chips.

**Comment by ajithme** [comment]
FYi This issue i found while using Apache Spark
I had created the original bug
https://github.com/scala/bug/issues/11429
due to 
https://issues.apache.org/jira/browse/SPARK-26961

**Comment by SethTisue** [comment]
> If adopted, this would probably need a community build

(happy to do that, someone tell me when it's ready)

**Comment by mkeskells** [comment]
> @mkeskells Did you come to this PR from a practical application performance issue or by discovering the problem via a review of this code?

It came from JFR observation while benchmarking and optimising a large application

> One problem I see with the implementation is that the approximation of the static initializer doesn't quite get the order correct. A subclass of `ScalaClassLoader.URLClassLoader` that attempts to register itself as parallel capable would have hard time guaranteeing that `ScalaClassLoader.URLClassLoader` had first been registered. It may be simpler to deprecate `ScalaClassLoader.URLClassLoader` and provide a replacement (which could be used internally by Toolbox evaluation etc) that is implemented as a Java class.

I did consider unconditionally initialising the class by placing a block in the outer object, but this is still messy. Its a shame that we cant have some low level annotations to specify methods to be @static or @staticInitialiser. That would fix this problem and a few others 😞

**Comment by retronym** [comment]
This test shows a problem with the proposed approach.

```scala
package scala.reflect

import org.junit.{Assert, Test}

import scala.reflect.internal.util.ScalaClassLoader

class ScalaClassLoaderTest {
  @Test
  def isRegisteredAsParallelCapable(): Unit = {
    val loader1, loader2 = ScalaClassLoader.fromURLs(Nil)
    checkIsParallelCabable(loader1)
    checkIsParallelCabable(loader2)
  }

  private def checkIsParallelCabable(loader: ClassLoader): Unit = {
    val fld = classOf[ClassLoader].getDeclaredField("parallelLockMap")
    fld.setAccessible(true)
    Assert.assertNotNull(fld.get(loader))
    // println(fld.get(loader) != null)
  }
}
```

The first classloader instance created will not be parallel capable because `java.net.URLClassLoader.<init>` (the super contructor) is called _before_ the registration, and that [inspects](https://github.com/openjdk/jdk/blob/8a9004da9b30c2ea3c616a0ef29be66f546faa78/src/java.base/share/classes/java/lang/ClassLoader.java#L379-L388) the registration status.

Subsequently created instances are parallel capable, although this is an implementation detail that contravenes the [spec](https://github.com/openjdk/jdk/blob/8a9004da9b30c2ea3c616a0ef29be66f546faa78/src/java.base/share/classes/java/lang/ClassLoader.java#L1597-L1603) of `registerAsParallelCabable`, which says that:

>  The registration succeeds if and only if all of the following conditions are met:
>   - no instance of the caller has been created
>   - all of the super classes (except class Object) of the caller are registered as parallel capable

So we should either switch to a Java implementation of the class or find a means to do register from within the static initializer.

Doing an "in-place" conversion of "ScalaClassLoader.URLClassLoader" to a java implementation looks hard, and maybe impossible as it is an inner class of `object ScalaClassLoader`. Perhaps we can have that enclosing object extends `/*java*/ class ScalaClassLoaderUtil { static class URLClassLoader { }`. I'll try that out shortly. If it doesn't work out, a fallback would be to introduce a brand new Java class and use that in the places that matter instead of `ScalaClassLoader.fromURLs`, notably from the `scala.tool.nsc.CommonRunner.run`

I've also experimented with an alternative that simulates a static inititalizer with a Java base class that calls into a static method in the Scala base class. We can generate a static method in Scala with a lambda. It is pretty hacky but I thought I'd note it here for posterity.


```diff
--- a/src/reflect/scala/reflect/internal/util/ScalaClassLoader.scala
+++ b/src/reflect/scala/reflect/internal/util/ScalaClassLoader.scala
@@ -130,9 +130,16 @@ object ScalaClassLoader {
   def setContext(cl: JClassLoader) = Thread.currentThread.setContextClassLoader(cl)

   class URLClassLoader(urls: Seq[URL], parent: JClassLoader)
-      extends JURLClassLoader(urls.toArray, parent)
+      extends URLClassLoaderStaticInit(urls.toArray, parent)
          with ScalaClassLoader
          with HasClassPath {
+    // Method name used by URLClassLoaderStaticInit.<static> to invoke the STATIC lambda implementation.
+    // `registerAsParallelCapable` is a caller-sensitive method and must be called from `URLClassLoader`,
+    // but it also needs to be called before we call the super constructor, i.e in the static initializer.
+    private def staticInit = () => {
+      ClassLoader.registerAsParallelCapable()
+    }

```

```java
package scala.reflect.internal.util;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.Arrays;

import static java.lang.reflect.Modifier.isStatic;

class URLClassLoaderStaticInit extends URLClassLoader {
  static {
    ClassLoader.registerAsParallelCapable();
    registerParallelCapable();
  }

  public URLClassLoaderStaticInit(URL[] urls, ClassLoader parent) {
    super(urls, parent);
  }

  private static void registerParallelCapable()  {
    try {
      Class<?> cls = Class.forName("scala.reflect.internal.util.ScalaClassLoader$URLClassLoader");
      for (Method declaredMethod : cls.getDeclaredMethods()) {
        if (isStatic(declaredMethod.getModifiers()) && declaredMethod.getName().startsWith("$anonfun$staticInit$")) {
          declaredMethod.setAccessible(true);
          declaredMethod.invoke(null);
        }
      }
    } catch (ClassNotFoundException | InvocationTargetException | IllegalAccessException e) {
      throw new ExceptionInInitializerError(e);
    }
  }
}

```

**Comment by retronym** [comment]
The details of introducing a Java-implemented, parallel capable replacement for `ScalaClassLoader.URLClassLoader` turned out to be rather awkward. This turned my attention to another approach -- just use `java.net.URLClassLoader` directly from the Scala script runner after refactoring some of the requisite reflection utility code to extension methods:

https://github.com/scala/scala/pull/9475

**Pull request closed without merging**